Add organizationSyncOptions prop to clerkMiddleware() options (#1634)
clerk/clerk-docs: docs/references/nextjs/clerk-middleware.mdx

  ---

  - `organizationSyncOptions?`
  - <code>[OrganizationSyncOptions](#organization-sync-options) | undefined</code>

  Used to activate a particular [organization](docs/organizations/overview) or the [personal account](/glossary#personal-account) based on path parameters in the URL.

  ---

  - `proxyUrl?`
  - `string`

  Specify the URL of the proxy, if using a proxy.

  ---

  - `publishableKey`
  - `string`

  The Clerk publishable key for your instance. This can be found in your Clerk Dashboard on the **[API Keys](https://dashboard.clerk.com/last-active?path=api-keys)** page.

  ---

  - `secretKey?`
  - `string`

  The Clerk secret key for your instance. This can be found in your Clerk Dashboard on the **[API Keys](https://dashboard.clerk.com/last-active?path=api-keys)** page. The `CLERK_ENCRYPTION_KEY` environment variable must be set when providing `secretKey` as an option, refer to [Dynamic keys](#dynamic-keys).

  ---

  - `signInUrl?`
  - `string`

  An alternative sign in URL.

  ---

  - `signUpUrl?`
  - `string`

>   An alternative sign up URL.

Author: This is all just a re-arrangement - I noticed that the list was _almost_ alphabetized, and now it's alphabetized.

  ---

  - `organizationSyncOptions?`
  - <code>[OrganizationSyncOptions](#organization-sync-options) | undefined</code>

>   Used to activate a specific [organization](docs/organizations/overview) or [personal account](https://clerk.com/glossary#personal-account) based on URL path parameters. If there's a mismatch between the active organization in the session (e.g., as reported by [`auth()`](/docs/references/nextjs/auth)) and the organization indicated by the URL, the middleware will attempt to activate the organization specified in the URL.

Author: Done! 90ad16deafd71d832c22cce78e5d48619bb3179c

Because the link doesn't hop directly to a personal account definition, I added a little text highlight to the link, so it will look like this:

![Screenshot 2024-10-29 at 9 27 58 AM](https://github.com/user-attachments/assets/0e1dbc37-46c0-4cca-903c-a068e896dc7d)

But I can go with the vanilla link if you'd prefer!

>   Used to activate a specific [organization](docs/organizations/overview) or [personal account](https://clerk.com/glossary#personal-account) based on URL path parameters. If there's a mismatch between the active organization in the session (e.g., as reported by [`auth()`](/docs/references/nextjs/auth)) and the organization indicated by the URL, the middleware will attempt to activate the organization specified in the URL.

Reviewer: ```suggestion
  Used to activate a specific [organization](docs/organizations/overview) or [personal account](/docs/organizations/organization-workspaces#organization-workspaces-in-the-clerk-dashboard:~:text=Personal%20account) based on URL path parameters. If there's a mismatch between the active organization in the session (e.g., as reported by [`auth()`](/docs/references/nextjs/auth)) and the organization indicated by the URL, the middleware will attempt to activate the organization specified in the URL.
```

  ---

  - `proxyUrl?`
  - `string`

  Specify the URL of the proxy, if using a proxy.

  ---

  - `publishableKey`
  - `string`

  The Clerk publishable key for your instance. This can be found in your Clerk Dashboard on the [**API Keys**](https://dashboard.clerk.com/last-active?path=api-keys) page.

  ---

  - `secretKey?`
  - `string`

>   The Clerk secret key for your instance. This can be found in your Clerk Dashboard on the **[API Keys](https://dashboard.clerk.com/last-active?path=api-keys)** page. The `CLERK_ENCRYPTION_KEY` environment variable must be set when providing `secretKey` as an option. For more information, refer to [Dynamic keys](#dynamic-keys) section.

Reviewer: ```suggestion
  The Clerk secret key for your instance. This can be found in your Clerk Dashboard on the **[API Keys](https://dashboard.clerk.com/last-active?path=api-keys)** page. The `CLERK_ENCRYPTION_KEY` environment variable must be set when providing `secretKey` as an option. For more information, refer to the [Dynamic keys](#dynamic-keys) section.
```

> [!NOTE]
> Dynamic keys are not accessible on the client-side.

### OrganizationSyncOptions

The `organizationSyncOptions` property on the <code>[clerkMiddleware()](#clerk-middleware-options)</code> options
object has the type `OrganizationSyncOptions`.

This defines the options for automatically activating an organization or the [personal account](/glossary#personal-account)
based on the URL. This is useful for applications that require organization slugs or IDs to appear in URL paths.

If a mismatch between the active organization on the session (e.g. as reported by
<code>[auth()](/docs/references/nextjs/auth)</code>) and the organization as indicated by the URL is detected, an attempt
to activate the given organization will be made.

`OrganizationSyncOptions` has the following properties:

<Properties>
  - `organizationPatterns`
  - <code>[Pattern](#pattern)\[]</code>

  URL patterns that are organization-specific and contain an organization ID or slug as a path parameter. If a request
  matches this path, the organization identifier will be extracted and activated before rendering.

>   If the route also matches the personalAccountPatterns, this takes precedence.

Reviewer 2: 🔧 Can we have `personalAccountPatterns` appear with in-line code formatting, or even link to the definition?

Author: 👍 

I don't see a precedent or a way to link to properties, but I can do the code formatting!

ff0d75ef3631d2f2665bd9508eb9af4a00e5a2db

>   Patterns must have a path parameter named either `:id` (matches a clerk organization ID) or `:slug` (matches a clerk

Reviewer 2: ```suggestion
  Patterns must have a path parameter named either `:id` (matches a Clerk organization ID) or `:slug` (matches a Clerk
```

  organization slug).

  > [!WARNING]
  > If the organization cannot be activated, either because it does not exist or the user lacks access,
  > no change will be made to the previously active organization. Components must detect this case and respond with an
  > appropriate error (e. g., `notFound()`, an

>   > [organization switcher](/docs/components/organization/organization-switcher), etc.).

Reviewer 2: 🔧 This sentence and examples in parentheses was confusing to me, because responding by rendering the org switcher doesn't seem to fall into the category of an appropriate _error_

Author: 73ce834fcd8913bc2cd41ef67ec313a4ac7a2f16

How does that feel?

  Common examples:
  - `["/orgs/:slug", "/orgs/:slug/(.*)"]`
  - `["/orgs/:id", "/orgs/:id/(.*)"]`

>   - `["/app/:any/orgs/:slug", "/app/:any/orgs/:slug/(.*)"]`

Reviewer 2: ❓ Would it be helpful to provide some examples that would match these?

Reviewer 2: Oh I just saw the section on Patterns, so maybe this isn't as necessary

> [!NOTE]
> Dynamic keys are not accessible on the client-side.

### `OrganizationSyncOptions`

The `organizationSyncOptions` property on the [`clerkMiddleware()`](#clerk-middleware-options) options
object has the type `OrganizationSyncOptions`, which has the following properties:

<Properties>
  - `organizationPatterns`
  - <code>[Pattern](#pattern)\[]</code>

  Specifies URL patterns that are organization-specific, containing an organization ID or slug as a path parameter. If a request

>   matches this path, the organization identifier will be extracted and activated before rendering.

Reviewer: when you say "the organization identifier will be activated", do you mean the organization identifier will be used to set that org as active?

Author: Yes, and I like your wording better!

e26600da08e4000e86654a273c62e36e14dd95c1

  If the route also matches the `personalAccountPatterns` prop, this prop takes precedence.

  Patterns must have a path parameter named either `:id` (to match a Clerk organization ID) or `:slug` (to match a Clerk organization slug).

  > [!WARNING]

>   > If the organization can't be activated—either because it doesn't exist or the user lacks access—the previously active organization will remain unchanged. Components must detect this case and provide an appropriate error and/or resolution pathway, such as calling `notFound()` or displaying an [organization switcher](/docs/components/organization/organization-switcher).

Reviewer: ```suggestion
  > If the organization can't be activated—either because it doesn't exist or the user lacks access—the previously active organization will remain unchanged. Components must detect this case and provide an appropriate error and/or resolution pathway, such as calling `notFound()` or displaying an [`<OrganizationSwitcher />`](/docs/components/organization/organization-switcher).
```

  Common examples:

  - `["/orgs/:slug", "/orgs/:slug/(.*)"]`
  - `["/orgs/:id", "/orgs/:id/(.*)"]`
  - `["/app/:any/orgs/:slug", "/app/:any/orgs/:slug/(.*)"]`

  ---

  - `personalAccountPatterns`
  - <code>[Pattern](#pattern)\[]</code>

>   URL patterns for resources that exist within the context of a [Clerk Personal Account](https://clerk.com/glossary#personal-account) (user-specific, outside any

Reviewer: ```suggestion
  URL patterns for resources that exist within the context of a [Clerk Personal Account](/docs/organizations/organization-workspaces#organization-workspaces-in-the-clerk-dashboard:~:text=Personal%20account) (user-specific, outside any
```

> [!NOTE]
> Dynamic keys are not accessible on the client-side.

### `OrganizationSyncOptions`

The `organizationSyncOptions` property on the [`clerkMiddleware()`](#clerk-middleware-options) options
object has the type `OrganizationSyncOptions`, which has the following properties:

<Properties>
  - `organizationPatterns`
  - <code>[Pattern](#pattern)\[]</code>

  URL patterns that are organization-specific and contain an organization ID or slug as a path parameter. If a request
  matches this path, the organization identifier will be extracted and activated before rendering.

  If the route also matches the `personalAccountPatterns` prop, this prop takes precedence.

  Patterns must have a path parameter named either `:id` (to match a Clerk organization ID) or `:slug` (to match a Clerk organization slug).

  > [!WARNING]
  > If the organization cannot be activated, either because it doesn't exist or the user lacks access,
  > no change will be made to the previously active organization. Components must detect this case and respond with an
  > appropriate error and/or resolution pathway (e.g., `notFound()` and an

>   > [organization switcher](/docs/components/organization/organization-switcher), etc.).

Author: Nice em-dashes! Do you have a trick for inserting the `—` literal, or do you copy-paste it from somewhere?

Question: why prefer to have the whole warning on one line, vs multiple lines?

  Common examples:

  - `["/orgs/:slug", "/orgs/:slug/(.*)"]`
  - `["/orgs/:id", "/orgs/:id/(.*)"]`
  - `["/app/:any/orgs/:slug", "/app/:any/orgs/:slug/(.*)"]`

  ---

  - `personalAccountPatterns`
  - <code>[Pattern](#pattern)\[]</code>

  URL patterns for resources that exist within the context of a [Clerk Personal Account](https://clerk.com/glossary#personal-account) (user-specific, outside any

>   organization).

Author: Done! https://github.com/clerk/clerk-docs/commit/90ad16deafd71d832c22cce78e5d48619bb3179c
